source/common/config/config_provider_impl.cc

```diff
@@ -0,0 +1,64 @@
+#include "common/config/config_provider_impl.h"
+
+namespace Envoy {
+namespace Config {
+
+StaticConfigProviderImpl::StaticConfigProviderImpl(
+    Server::Configuration::FactoryContext& factory_context,
+    ConfigProviderManagerImpl& config_provider_manager)
+    : last_updated_(factory_context.timeSource().systemTime()),
+      config_provider_manager_(config_provider_manager) {
+  config_provider_manager_.static_config_providers_.insert(this);
+}
+
+StaticConfigProviderImpl::~StaticConfigProviderImpl() {
+  config_provider_manager_.static_config_providers_.erase(this);
+}
+
+ConfigSubscriptionInstance::~ConfigSubscriptionInstance() {
+  runInitializeCallbackIfAny();
+  config_provider_manager_.unbindSubscription(manager_identifier_);
+}
+
+void ConfigSubscriptionInstance::runInitializeCallbackIfAny() {
+  if (initialize_callback_) {
+    initialize_callback_();
+    initialize_callback_ = nullptr;
+  }
+}
+
+bool ConfigSubscriptionInstance::checkAndApplyConfig(const Protobuf::Message& config_proto,
+                                                     const std::string& config_name,
+                                                     const std::string& version_info) {
+  const uint64_t new_hash = MessageUtil::hash(config_proto);
+  if (config_info_ && config_info_.value().last_config_hash_ == new_hash) {
+    return false;
+  }
+
+  config_info_ = {new_hash, version_info};
+  ENVOY_LOG(debug, "{}: loading new configuration: config_name={} hash={}", name_, config_name,
+            new_hash);
+
+  ASSERT(!dynamic_config_providers_.empty());
+  ConfigProvider::ConfigConstSharedPtr new_config;
+  for (auto* provider : dynamic_config_providers_) {
```

Comment:
I'm scratching my head and trying to remember why we have/need multiple dynamic config providers for a single config proto update?

Reply:
This is another carryover from the existing RDS implementation of config providers. Currently, each HttpConnectionManager [instantiates its own RdsRouteConfigProviderImpl](https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/network/http_connection_manager/config.cc#L173) (if RDS is configured). 

I found some context on the decision to avoid sharing the RouteConfigProvider in #3960 and #3967. Based on those PRs, it appears the goal was to avoid use of shared_ptr outside the *ProviderImpl.

Reply:
Added class level comment explaining this.

Reply:
> The goal was to avoid use of shared_ptr outside the *ProviderImpl.

This is not quite true, see #3960 description. FactoryContext is tied to Listener (some of refs in it), and they will be dangling after they are updated by LDS, while the config provider may outlive them.